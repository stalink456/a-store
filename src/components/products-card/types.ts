export type ProductsCardType = {
  id: number;
  preview: string;
  title: string;
  subtitle?: string;
  price: number;
};
