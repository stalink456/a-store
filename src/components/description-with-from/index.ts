export { DescriptionWithForm } from './description-with-form';
