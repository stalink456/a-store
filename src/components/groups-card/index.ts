export { GroupsCard } from './groups-card';
