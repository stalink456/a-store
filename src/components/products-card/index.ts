export { ProductsCard } from "./products-card";
