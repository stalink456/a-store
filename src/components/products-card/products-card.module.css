@import "../../css/variables.css";
@import "../../css/animation.css";

.products_card {
  position: relative;
  width: 340px;
  padding-bottom: 75px;
  padding: 0 7px;
  margin-top: 20px;
  composes: slide_right from global;
}

.products_card p {
  margin-bottom: 5px;
}
.products_card a {
  color: inherit;
  text-decoration: inherit;
}

.products_card img {
  max-height: 350px;
  max-width: 100%;
  object-fit: cover;
  object-position: top 20%;
}

.products_card img:hover {
  filter: brightness(80%);
}

@media (max-width: 1065px) {
  .products_card {
    width: 305px;
    margin-top: 20px;
  }
  .products_card img {
    max-width: 100%;
  }
}

@media (max-width: 660px) {
  .products_card {
    width: 235px;
  }

  .products_card h6 {
    font-size: 16px;
  }

  .products_card p {
    font-size: 14px;
  }
}

@media (max-width: 520px) {
  .products_card {
    width: 210px;
  }
}

@media (max-width: 480px) {
  .products_card {
    width: 370px;
    padding: 0;
  }
}
