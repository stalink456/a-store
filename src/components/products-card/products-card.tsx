import React from "react";
import { Link } from "react-router-dom";
import { Space } from "@alfalab/core-components/space";
import { Typography } from "@alfalab/core-components/typography";
import { LoadableImage } from "components/ui-components/loadable-image";
import { ProductsCardType } from "./types";

import styles from "./products-card.module.css";

export const ProductsCard: React.FC<ProductsCardType> = ({
  id,
  preview,
  title,
  subtitle,
  price,
}) => {
  return (
    <div className={styles.products_card}>
      <Link to={`/product/${id}`}>
        <Space direction="vertical" size={0} wrap>
          <Space direction="horizontal">
            <LoadableImage src={preview} alt={title} />
          </Space>
          <Space direction="horizontal">
            <Typography.Title tag="h6" view="xsmall" weight="medium">
              {title}
            </Typography.Title>
          </Space>
          {subtitle ? (
            <Space direction="horizontal">
              <Typography.Text
                tag="p"
                view="primary-medium"
                weight="bold"
                color="tertiary-inverted"
                data-testid={`subtitle-${id}`}
              >
                {subtitle}
              </Typography.Text>
            </Space>
          ) : null}
          <Space direction="horizontal" size={0}>
            <Typography.Text
              tag="p"
              weight="bold"
              view="primary-large"
              data-testid={`price-${id}`}
            >
              {price} â‚½
            </Typography.Text>
          </Space>
        </Space>
      </Link>
    </div>
  );
};
