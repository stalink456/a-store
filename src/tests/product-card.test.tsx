import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { ProductsCard } from 'components/products-card';
import Product from 'pages/product';
import * as product from './mocks/get-product-200.json';
import * as cardItem from './mocks/get-card-item-200.json';
import * as loading from './mocks/get-loading-false.json';

const mockedProduct = product['product'];
const mockedCard = cardItem['card'];
const mockIsLoading = loading['loading'];

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('hooks/use-product', () => ({
  useProduct: () => ({
    product: mockedProduct,
    isLoading: mockIsLoading,
  }),
}));

describe('ProductsCard component', () => {
  it('should render ProductsCard component', () => {
    const { baseElement } = render(<ProductsCard {...mockedCard} />, {
      wrapper: MemoryRouter,
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('should render Menu component with a transition to ProductsCard page', async () => {
    render(
      <MemoryRouter>
        <ProductsCard {...mockedCard} />
        <Routes>
          <Route path='/' element={<></>} />
          <Route path='/product/:id' element={<Product />} />
        </Routes>
      </MemoryRouter>
    );

    fireEvent.click(screen.getByAltText('Блокнот Для умных и свободных'));
  });

  it('should render ProductsCard component with image', async () => {
    render(<ProductsCard {...mockedCard} />, { wrapper: MemoryRouter });

    const image = (await screen.findByAltText(
      'Блокнот Для умных и свободных'
    )) as HTMLImageElement;
    await waitFor(() => expect(image.src !== '').toBeTruthy());
  });

  it('should render ProductsCard component with description', async () => {
    render(<ProductsCard {...mockedCard} />, { wrapper: MemoryRouter });

    const result = await screen.findByText('Блокнот Для умных и свободных');
    expect(result).toBeTruthy();
  });

  it("shouldn't render ProductsCard component with non-existent price", () => {
    render(<ProductsCard {...mockedCard} />, { wrapper: MemoryRouter });
    const price = screen.getByTestId(`price-${mockedCard.id}`).textContent;

    expect(price).not.toBe('1999 ₽');
  });
});
