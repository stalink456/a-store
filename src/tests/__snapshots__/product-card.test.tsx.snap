// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsCard component should render ProductsCard component 1`] = `
<body>
  <div>
    <div
      class="products_card"
    >
      <a
        href="/product/2"
      >
        <div
          class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as"
          style="flex-wrap: wrap; margin-bottom: 0px;"
        >
          <div
            class="space__spaceItem_nr3as"
            style="margin-bottom: 0px;"
          >
            <div
              class="space__spaceContainer_nr3as space__start_nr3as"
            >
              <div
                class="space__spaceItem_nr3as"
              >
                <div
                  class="skeleton__component_1bshy skeleton__animate_1bshy"
                >
                  <img
                    alt="Блокнот Для умных и свободных"
                    class=""
                    src="http://qa-games.ru/astore/public/images/77117755.jpeg"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="space__spaceItem_nr3as"
            style="margin-bottom: 0px;"
          >
            <div
              class="space__spaceContainer_nr3as space__start_nr3as"
            >
              <div
                class="space__spaceItem_nr3as"
              >
                <h6
                  class="typography__component_19zgr typography__styrene-xsmall_pxbut typography__medium_pxbut"
                >
                  Блокнот Для умных и свободных
                </h6>
              </div>
            </div>
          </div>
          <div
            class="space__spaceItem_nr3as"
          >
            <div
              class="space__spaceContainer_nr3as space__start_nr3as"
            >
              <div
                class="space__spaceItem_nr3as"
              >
                <p
                  class="typography__paragraph_c16y5 typography__primary-large_c16y5 typography__bold_c16y5"
                  data-testid="price-2"
                >
                  1499
                   ₽
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</body>
`;
